firebase-perf/src/main/java/com/google/firebase/perf/util/Timer.java

```diff
@@ -14,92 +14,130 @@
    */
-  private long highResTime;
+  private static long wallClockMicros() {
+    return MILLISECONDS.toMicros(System.currentTimeMillis());
+  }
 
   /**
-   * Construct Timer object using System clock. Make it package visible to be only accessible from
-   * com.google.firebase.perf.util.Clock.
+   * Helper to get current {@link SystemClock#elapsedRealtime()} from system API.
+   *
+   * @return wall-clock time in microseconds.
    */
+  private static long elapsedRealtimeMicros() {
+    return MILLISECONDS.toMicros(SystemClock.elapsedRealtime());
+  }
+
+  // TODO: make all constructors private, use public static factory methods, per Effective Java
+  /** Construct Timer object using System clock. */
   public Timer() {
-    timeInMicros = TimeUnit.MILLISECONDS.toMicros(System.currentTimeMillis());
-    highResTime = System.nanoTime();
+    this(wallClockMicros(), elapsedRealtimeMicros());
   }
 
   /**
-   * Construct a Timer object with input wall-clock time, assume high resolution time is same as
-   * wall-clock time.
+   * Construct a Timer object with input wall-clock time and elapsedRealtime.
    *
-   * @param time wall-clock time in microseconds
+   * @param epochMicros wall-clock time in milliseconds since epoch
+   * @param elapsedRealtimeMicros monotonic time in microseconds in the {@link
+   *     SystemClock#elapsedRealtime()} timebase
    */
   @VisibleForTesting
-  public Timer(long time) {
-    this.timeInMicros = time;
-    highResTime = TimeUnit.MICROSECONDS.toNanos(time);
+  Timer(long epochMicros, long elapsedRealtimeMicros) {
+    this.wallClockMicros = epochMicros;
+    this.elapsedRealtimeMicros = elapsedRealtimeMicros;
   }
 
   /**
-   * Construct a Timer object with input wall-clock time and high resolution time.
+   * TEST-ONLY constructor that sets both wall-clock time and elapsedRealtime to the same input
+   * value. Do NOT use this for any real logic because this is mixing 2 different time-bases.
    *
-   * @param time wall-clock time in microseconds
-   * @param highResTime high resolution time in nanoseconds
+   * @param testTime value to set both wall-clock and elapsedRealtime to for testing purposes
    */
   @VisibleForTesting
-  public Timer(long time, long highResTime) {
-    this.timeInMicros = time;
-    this.highResTime = highResTime;
+  public Timer(long testTime) {
+    this(testTime, testTime);
   }
 
   private Timer(Parcel in) {
-    timeInMicros = in.readLong();
-    highResTime = in.readLong();
+    this(in.readLong(), in.readLong());
   }
 
   /** resets the start time */
   public void reset() {
-    timeInMicros = TimeUnit.MILLISECONDS.toMicros(System.currentTimeMillis());
-    highResTime = System.nanoTime();
+    // TODO: consider removing this method and make Timer immutable thus fully thread-safe
+    wallClockMicros = wallClockMicros();
+    elapsedRealtimeMicros = elapsedRealtimeMicros();
   }
 
   /** Return wall-clock time in microseconds. */
   public long getMicros() {
```

Comment:
+1

Reply:
For `getMicros`, I don't know the original author's thoughts exactly but I'd imagine it's because they wanted to encapsulate "wall-clock" and "elapsedRealtime" so anyone who uses `Timer` don't need to understand them. Unfortunately they added `getElapsedRealtimeMicros()` for testing purposes which makes `getMicros` confusing and defeats the whole purpose of encapsulation. 

I'm not against changing the name because I also like more explicit names, but I also propose that we remove `getElapsedRealtimeMicros()` because it's test-only.

Reply:
> How hard/tricky is to refactor them?

Definitely not hard, just that these touch a lot of files and I don't feel like it makes sense to make a feature-implementation PR heavy with other things. I feel like this PR should be focused on getting the `elapsedRealtime`-change right, and I would like to do other code health improvements in a separate PR.